tests/CSSList/DocumentTest.php

```diff
@@ -85,4 +85,66 @@ public function setContentsReplacesContentsSetInPreviousCall()
 
         self::assertSame($contents2, $this->subject->getContents());
     }
+
+    /**
+     * @return \Generator<array<array<DeclarationBlock>, DeclarationBlock, DeclarationBlock, array<DeclarationBlock>>>
+     */
+    public static function insertBeforeDataProvider(): \Generator
+    {
+
+        $bogusOne = new DeclarationBlock();
+        $bogusOne->setSelectors('.bogus-one');
+        $bogusTwo = new DeclarationBlock();
+        $bogusTwo->setSelectors('.bogus-two');
+
+        $oItem = new DeclarationBlock();
+        $oItem->setSelectors('.item');
+
+        $oSibling = new DeclarationBlock();
+        $oSibling->setSelectors('.sibling');
+
+        $oOrphan = new DeclarationBlock();
+        $oOrphan->setSelectors('.forever-alone');
+
+        yield 'insert before' => [
+            'initialContent' => [$bogusOne, $oSibling, $bogusTwo],
+            'oItem' => $oItem,
+            'oSibling' => $oSibling,
+            'expectedContent' => [$bogusOne, $oItem, $oSibling, $bogusTwo],
+            ];
+        yield 'append if not found' => [
+            'initialContent' => [$bogusOne, $oSibling, $bogusTwo],
+            'oItem' => $oItem,
+            'oSibling' => $oOrphan,
+            'expectedContent' => [$bogusOne, $oSibling, $bogusTwo, $oItem],
+            ];
+    }
+
+    /**
+     * @test
+     *
+     * @param $initialContent
+     * @param $oItem
+     * @param $oSibling
+     * @param $sPosition
+     * @param $expectedContent
+     *
+     * @dataProvider insertBeforeDataProvider
+     */
+    public function insertContentBefore(
```

Comment:
The tested method has two kinds of behavior: replacing and inserting. Please let's have two different tests for this.

The test name should be a statement communicating the tested behavior. This allows the test name and test code to form a kind of "double bookkeeping". Please have a look at this slide about naming tests: https://speakerdeck.com/oliverklee/test-driven-development-with-phpunit-915a5f2c-3d37-4e41-b5bc-ef1efca9c01e?slide=39

Reply:
I tend to dissagree. As the method name suggests, we insert an item. Yes, it does this by internally mapping to the existing `replace()` function, but in the end, for the consumer the item is inserted and not replaced. It does append (inserts at the end, before the not found sibbling), if it can't find the spot to insert the item, but this two different insertion points are reflected in the naming of the test data in the dataprovider.

Splitting this in to two test methods basiaclly doublicates the code, because the test code would be exactly the same.

But if you insist, I will make the change.

Reply:
Yes, please. (I find some duplication in tests okay as long as it helps them tell an easy-to read story.)

Reply:
We probably then also can do without the data providers and instead include the data directly in the corresponding tests.

Reply:
I separated the two tests, as requested. Happy now?

Reply:
But where's the point of having data providers if you separate everything into its own testing method? This is now a lot of duplicate code. Take e.g. `expandBorderShorthand()` and its `expandBorderShorthandProvider()` in `tests/RuleSet/DeclarationBlockTest.php`. Shouldn't this rather be:

- `expandBorderShorthandExpandsBorderWithBorderStyleAndBorderColor`
- `expandBorderShorthandExpandsBorderStyle`
- `expandBorderShorthandExpandsBorderWidth`
- `expandBorderShorthandExpandsBorderColor`
- `expandBorderShorthandExpandsBorderWithBorderStyle`
- `expandBorderShorthandFixesWhitespace`

following the same approach as done here?

Reply:
For testing `expandBorderShorthand`, I'd differentiate between two pieces of behavior:

1. that is expands different border things (I'd use a data provider for that)
2. that it fixes whitespace

So concerning splitting tests, I'd differentiate between "the same behavior, but with different data" and different behavior.

Let's have a non-code example - maybe this will help me communicate my thinking. :-)

- different behavior: I can eat. I can draw. I can spit things out.
- same behavior, but with different data (data providers): I can eat apples, bananas and bread.

Does this help, @ziegenberg?

Would it be helpful to have a call to talk about testing strategies and patterns?